src/main/scala/org/constellation/primitives/concurrency/SingleLock.scala

```diff
@@ -0,0 +1,21 @@
+package org.constellation.primitives.concurrency
+import cats.effect.concurrent.Semaphore
+import cats.effect.{Concurrent, Timer}
+import cats.implicits._
+
+import scala.concurrent.duration._
+
+class SingleLock[F[_], R](name: String, s: Semaphore[F], op: => F[R])(implicit F: Concurrent[F]){
+  def acquire: F[R] =
+    for {
+      x <- s.available
+      _ <- F.delay(println(s"$name >> Availability: $x"))
+      _ <- s.acquire
+      y <- s.available
+      _ <- F.delay(println(s"$name >> Started | Availability: $y"))
+      res <- op
```

Comment:
I'm not 100% sure -- so I'd recommend testing this -- but I believe that if op returns an error, most implementations of F (cats IO, for example) are going to short-circuit the rest of these commands. Meaning that your semaphore won't get released, and all hell will break loose. 

Again, let's test this to be sure, but I think you'd want to use something like bracket (https://typelevel.org/cats-effect/typeclasses/bracket.html) here.

Reply:
Oh cool thanks for the tip!